File: experiments/engine-kt/tools/src/main/kotlin/com/garive/eng/kt/tools/GovernedReducer.kt
package com.garive.eng.kt.tools

/** Durable authorization verdict for one exact invocation. */
public sealed interface AuthorizationVerdict {
    /** Exact authority grant. */
    public data class Approve(public val grant: InvocationGrant) : AuthorizationVerdict
    /** Stable safe denial. */
    public data class Deny(public val code: String, public val details: String?) : AuthorizationVerdict
    /** Original invocation terminates without authority. */
    public data object ReplacementRequired : AuthorizationVerdict
    /** Interaction must suspend the current Execution. */
    public data class InteractionRequired(public val request: InteractionRequest) : AuthorizationVerdict
}

/** Suspension terminal for the current Execution. */
public sealed interface SuspensionRequirement {
    /** Committed interaction awaits continuation. */
    public data class Interaction(public val request: InteractionRequest) : SuspensionRequirement
    /** Uncertain effect requires operator reconciliation. */
    public data class OperatorReconciliation(public val evidence: String) : SuspensionRequirement
}

/** Next action after one committed portable input. */
public sealed interface GovernedAction {
    /** Ask Runtime authorization again. */
    public data object Authorize : GovernedAction
    /** Ask Runtime execution for the exact grant. */
    public data class Dispatch(public val grant: InvocationGrant) : GovernedAction
    /** Return safe model-visible feedback. */
    public data class Observation(public val observation: GovernedObservation) : GovernedAction
    /** Suspend the current Execution. */
    public data class Suspend(public val requirement: SuspensionRequirement) : GovernedAction
    /** Fail closed. */
    public data class Fail(public val code: GovernedFailureCode) : GovernedAction
    /** Idempotent duplicate produced no action. */
    public data object None : GovernedAction
}

private sealed interface ReducerState {
    data object Prepared : ReducerState
    data class Awaiting(val request: InteractionRequest) : ReducerState
    data class Authorized(val grant: InvocationGrant) : ReducerState
    data class Started(val grant: InvocationGrant) : ReducerState
    data object Denied : ReducerState
    data object Replaced : ReducerState
    data object Completed : ReducerState
    data object Failed : ReducerState
    data object Uncertain : ReducerState
}

/** One sequential governed invocation reducer carrying no Runtime I/O. */
public class GovernedEffect private constructor(
    private val invocationId: ToolInvocationId,
    private val prepared: PreparedToolCall,
) {
    private var reducerState: ReducerState = ReducerState.Prepared
    private var lastInteractionResolution: InteractionResolution? = null

    /** Current portable lifecycle state. */
    public val state: EffectState
        get() = when (reducerState) {
            ReducerState.Prepared -> EffectState.PREPARED
            is ReducerState.Awaiting -> EffectState.AWAITING_INTERACTION
            is ReducerState.Authorized -> EffectState.AUTHORIZED
            is ReducerState.Started -> EffectState.STARTED
            ReducerState.Denied -> EffectState.DENIED
            ReducerState.Replaced -> EffectState.REPLACED
            ReducerState.Completed -> EffectState.COMPLETED
            ReducerState.Failed -> EffectState.FAILED
            ReducerState.Uncertain -> EffectState.UNCERTAIN
        }

    /** Reduces one durably committed authorization verdict. */
    public fun applyAuthorization(verdict: AuthorizationVerdict): GovernedAction {
        val current = reducerState
        if (current is ReducerState.Authorized && verdict is AuthorizationVerdict.Approve) {
            return if (current.grant == verdict.grant) GovernedAction.None else fail(GovernedFailureCode.INVOCATION_CONFLICT)
        }
        if (current != ReducerState.Prepared) return fail(GovernedFailureCode.INVOCATION_CONFLICT)
        return when (verdict) {
            is AuthorizationVerdict.Approve -> {
                if (!grantBinds(verdict.grant)) return fail(GovernedFailureCode.GRANT_MISMATCH)
                reducerState = ReducerState.Authorized(verdict.grant)
                GovernedAction.Dispatch(verdict.grant)
            }
            is AuthorizationVerdict.Deny -> {
                if (verdict.code.isEmpty()) return fail(GovernedFailureCode.INVOCATION_CONFLICT)
                reducerState = ReducerState.Denied
                GovernedAction.Observation(observation(ObservationOutcome.Rejected(verdict.code, verdict.details)))
            }
            AuthorizationVerdict.ReplacementRequired -> {
                reducerState = ReducerState.Replaced
                GovernedAction.Observation(observation(ObservationOutcome.Rejected("replacement_required", null)))
            }
            is AuthorizationVerdict.InteractionRequired -> {
                val request = verdict.request
                if (request.preparedDigest.isEmpty() || request.expiryPolicy.isEmpty() ||
                    PortableSchema.validateValueDefinition(request.responseSchema) != null ||
                    request.invocationId != invocationId || request.preparedDigest != prepared.inputDigest
                ) return fail(GovernedFailureCode.INTERACTION_CONFLICT)
                reducerState = ReducerState.Awaiting(request)
                lastInteractionResolution = null
                GovernedAction.Suspend(SuspensionRequirement.Interaction(request))
            }
        }
    }

    /** Reduces committed interaction continuation without inventing authority. */
    public fun applyInteraction(resolution: InteractionResolution): GovernedAction {
        lastInteractionResolution?.let { existing ->
            return if (existing == resolution) GovernedAction.None else fail(GovernedFailureCode.INTERACTION_CONFLICT)
        }
        val request = (reducerState as? ReducerState.Awaiting)?.request
            ?: return fail(GovernedFailureCode.INTERACTION_CONFLICT)
        val matches = when (resolution) {
            is InteractionResolution.Resolved -> resolution.interactionId == request.interactionId && resolution.invocationId == request.invocationId && resolution.preparedDigest == request.preparedDigest && PortableSchema.validateArguments(request.responseSchema, resolution.response).isEmpty()
            is InteractionResolution.Cancelled -> resolution.interactionId == request.interactionId && resolution.invocationId == request.invocationId && resolution.preparedDigest == request.preparedDigest
        }
        if (!matches) return fail(GovernedFailureCode.INTERACTION_CONFLICT)
        return when (resolution) {
            is InteractionResolution.Resolved -> { lastInteractionResolution = resolution; reducerState = ReducerState.Prepared; GovernedAction.Authorize }
            is InteractionResolution.Cancelled -> { lastInteractionResolution = resolution; reducerState = ReducerState.Denied; GovernedAction.Observation(observation(ObservationOutcome.Rejected("interaction_cancelled", null))) }
        }
    }

    /** Reduces one Runtime execution fact after its durable commit. */
    public fun applyExecution(fact: ExecutionFact): GovernedAction {
        val current = reducerState
        return when {
            current is ReducerState.Authorized && fact is ExecutionFact.Started -> {
                reducerState = ReducerState.Started(current.grant); GovernedAction.None
            }
            current is ReducerState.Authorized && fact is ExecutionFact.Unsupported -> {
                if (fact.requirement.isEmpty()) fail(GovernedFailureCode.CORRUPT_RECOVERY_STATE) else fail(GovernedFailureCode.REQUIREMENT_UNSUPPORTED)
            }
            current is ReducerState.Started && fact is ExecutionFact.Completed -> {
                val receipt = fact.receipt ?: return fail(GovernedFailureCode.CORRUPT_RECOVERY_STATE)
                if (!receiptBinds(receipt, current.grant, TerminalClassification.COMPLETED) || fact.content.toString().encodeToByteArray().size.toLong() > current.grant.grantedRequirements.maxOutputBytes) return fail(GovernedFailureCode.CORRUPT_RECOVERY_STATE)
                reducerState = ReducerState.Completed
                GovernedAction.Observation(observation(ObservationOutcome.Succeeded(fact.content, fact.truncated)))
            }
            current is ReducerState.Started && fact is ExecutionFact.Failed -> {
                val receipt = fact.receipt ?: return fail(GovernedFailureCode.CORRUPT_RECOVERY_STATE)
                if (fact.code.isEmpty() || !receiptBinds(receipt, current.grant, TerminalClassification.FAILED) || fact.partial?.toString()?.encodeToByteArray()?.size?.toLong()?.let { it > current.grant.grantedRequirements.maxOutputBytes } == true) return fail(GovernedFailureCode.CORRUPT_RECOVERY_STATE)
                reducerState = ReducerState.Failed
                GovernedAction.Observation(observation(ObservationOutcome.Failed(fact.code, fact.details, fact.partial)))
            }
            current is ReducerState.Started && fact is ExecutionFact.Uncertain -> {
                if (fact.evidence.isEmpty()) return fail(GovernedFailureCode.CORRUPT_RECOVERY_STATE)
                reducerState = ReducerState.Uncertain
                GovernedAction.Suspend(SuspensionRequirement.OperatorReconciliation(fact.evidence))
            }
            else -> fail(GovernedFailureCode.INVOCATION_CONFLICT)
        }
    }

    private fun grantBinds(grant: InvocationGrant): Boolean =
        grant.constraintsDigest.isNotEmpty() && grant.authorityRevision.isNotEmpty() && grant.invocationId == invocationId && grant.preparedDigest == prepared.inputDigest && grant.toolName == prepared.toolName && grant.toolRevision == prepared.toolRevision && grant.grantedRequirements.maxDurationMs <= prepared.requirements.maxDurationMs && grant.grantedRequirements.maxOutputBytes <= prepared.requirements.maxOutputBytes && prepared.requirements.capabilities.containsAll(grant.grantedRequirements.capabilities)

    private fun receiptBinds(receipt: EffectReceipt, grant: InvocationGrant, terminal: TerminalClassification): Boolean =
        receipt.preparedDigest.isNotEmpty() && receipt.executorId.isNotEmpty() && receipt.executorRevision.isNotEmpty() && receipt.resultDigest.isNotEmpty() && receipt.invocationId == invocationId && receipt.preparedDigest == prepared.inputDigest && receipt.grantId == grant.grantId && receipt.terminalClassification == terminal

    private fun observation(outcome: ObservationOutcome): GovernedObservation =
        GovernedObservation(invocationId, prepared.inputDigest, prepared.modelCallId, prepared.toolName, outcome)

    private fun fail(code: GovernedFailureCode): GovernedAction {
        reducerState = ReducerState.Failed
        return GovernedAction.Fail(code)
    }

    public companion object {
        /** Starts at Prepared and asks Runtime for authorization. */
        public fun start(invocationId: ToolInvocationId, prepared: PreparedToolCall): Pair<GovernedEffect, GovernedAction> =
            GovernedEffect(invocationId, prepared) to GovernedAction.Authorize
    }
}

/** Selects recovery without treating a replay declaration as executor proof. */
public fun recoverEffect(position: RecoveryPosition, replayClass: ReplayClass, executorProvesReplay: Boolean): RecoveryDecision =
    when (position) {
        RecoveryPosition.AUTHORIZED -> RecoveryDecision.REVALIDATE_GRANT
        RecoveryPosition.RECEIPT_NO_RESULT -> RecoveryDecision.RECONSTRUCT_FROM_RECEIPT
        RecoveryPosition.TERMINAL -> RecoveryDecision.RETURN_TERMINAL
        RecoveryPosition.STARTED_NO_RECEIPT -> when {
            replayClass == ReplayClass.RECEIPT_RECOVERABLE && executorProvesReplay -> RecoveryDecision.RECOVER_EXECUTOR_RECEIPT
            replayClass in setOf(ReplayClass.READ_ONLY, ReplayClass.IDEMPOTENT) && executorProvesReplay -> RecoveryDecision.RETRY_SAME_INVOCATION
            else -> RecoveryDecision.RECONCILE_OPERATOR
        }
    }

File: engine/tools/src/governed_reducer.rs
//! Deterministic C5 authority, interaction, execution, and recovery reduction.

use crate::schema::{validate_arguments, validate_value_definition};
use crate::{
    EffectReceipt, EffectState, ExecutionFact, GovernedAction, GovernedEffectFailure,
    GovernedFailureCode, GovernedObservation, InteractionRequest, InteractionResolution,
    InvocationGrant, ObservationOutcome, PreparedToolCall, RecoveryDecision, RecoveryPosition,
    ReplayClass, SuspensionRequirement, TerminalClassification, ToolInvocationId,
};

/// Durable authorization verdict for the exact invocation and Prepared Call.
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum AuthorizationVerdict {
    /// Exact authority grant.
    Approve(InvocationGrant),
    /// Stable safe denial.
    Deny {
        /// Stable denial code.
        code: String,
        /// Optional redacted detail.
        details: Option<String>,
    },
    /// Original invocation terminates without executable authority.
    ReplacementRequired,
    /// A committed interaction must suspend the current Execution.
    InteractionRequired(InteractionRequest),
}

#[derive(Clone, Debug, Eq, PartialEq)]
enum ReducerState {
    Prepared,
    AwaitingInteraction(InteractionRequest),
    Authorized(InvocationGrant),
    Started(InvocationGrant),
    Denied,
    Replaced,
    Completed,
    Failed,
    Uncertain,
}

/// One sequential governed invocation reducer carrying no Runtime I/O.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct GovernedEffect {
    invocation_id: ToolInvocationId,
    prepared: PreparedToolCall,
    state: ReducerState,
    last_interaction_resolution: Option<InteractionResolution>,
}

impl GovernedEffect {
    /// Starts at Prepared and asks Runtime to authorize the exact call.
    pub fn new(
        invocation_id: ToolInvocationId,
        prepared: PreparedToolCall,
    ) -> (Self, GovernedAction) {
        (
            Self {
                invocation_id,
                prepared,
                state: ReducerState::Prepared,
                last_interaction_resolution: None,
            },
            GovernedAction::Authorize,
        )
    }

    /// Returns the portable lifecycle state.
    pub const fn state(&self) -> EffectState {
        match self.state {
            ReducerState::Prepared => EffectState::Prepared,
            ReducerState::AwaitingInteraction(_) => EffectState::AwaitingInteraction,
            ReducerState::Authorized(_) => EffectState::Authorized,
            ReducerState::Started(_) => EffectState::Started,
            ReducerState::Denied => EffectState::Denied,
            ReducerState::Replaced => EffectState::Replaced,
            ReducerState::Completed => EffectState::Completed,
            ReducerState::Failed => EffectState::Failed,
            ReducerState::Uncertain => EffectState::Uncertain,
        }
    }

    /// Reduces one durably committed authorization verdict.
    pub fn apply_authorization(&mut self, verdict: AuthorizationVerdict) -> GovernedAction {
        if let (ReducerState::Authorized(existing), AuthorizationVerdict::Approve(candidate)) =
            (&self.state, &verdict)
        {
            return if existing == candidate {
                GovernedAction::None
            } else {
                self.fail(GovernedFailureCode::InvocationConflict)
            };
        }
        if self.state != ReducerState::Prepared {
            return self.fail(GovernedFailureCode::InvocationConflict);
        }
        match verdict {
            AuthorizationVerdict::Approve(grant) => {
                if !self.grant_binds(&grant) {
                    return self.fail(GovernedFailureCode::GrantMismatch);
                }
                self.state = ReducerState::Authorized(grant.clone());
                GovernedAction::Dispatch(grant)
            }
            AuthorizationVerdict::Deny { code, details } => {
                if code.is_empty() {
                    return self.fail(GovernedFailureCode::InvocationConflict);
                }
                self.state = ReducerState::Denied;
                GovernedAction::Observation(
                    self.observation(ObservationOutcome::Rejected { code, details }),
                )
            }
            AuthorizationVerdict::ReplacementRequired => {
                self.state = ReducerState::Replaced;
                GovernedAction::Observation(self.observation(ObservationOutcome::Rejected {
                    code: "replacement_required".to_owned(),
                    details: None,
                }))
            }
            AuthorizationVerdict::InteractionRequired(request) => {
                if request.validate().is_err()
                    || validate_value_definition(&request.response_schema).is_err()
                    || request.invocation_id != self.invocation_id
                    || request.prepared_digest != self.prepared.input_digest()
                {
                    return self.fail(GovernedFailureCode::InteractionConflict);
                }
                self.state = ReducerState::AwaitingInteraction(request.clone());
                self.last_interaction_resolution = None;
                GovernedAction::Suspend(SuspensionRequirement::Interaction(request))
            }
        }
    }

    /// Reduces a committed interaction continuation without inventing authority.
    pub fn apply_interaction(&mut self, resolution: InteractionResolution) -> GovernedAction {
        if let Some(existing) = &self.last_interaction_resolution {
            return if existing == &resolution {
                GovernedAction::None
            } else {
                self.fail(GovernedFailureCode::InteractionConflict)
            };
        }
        let ReducerState::AwaitingInteraction(request) = &self.state else {
            return self.fail(GovernedFailureCode::InteractionConflict);
        };
        let matches = match &resolution {
            InteractionResolution::Resolved {
                interaction_id,
                invocation_id,
                prepared_digest,
                response,
            } => {
                interaction_id == &request.interaction_id
                    && invocation_id == &request.invocation_id
                    && prepared_digest == &request.prepared_digest
                    && validate_arguments(&request.response_schema, response).is_empty()
            }
            InteractionResolution::Cancelled {
                interaction_id,
                invocation_id,
                prepared_digest,
            } => {
                interaction_id == &request.interaction_id
                    && invocation_id == &request.invocation_id
                    && prepared_digest == &request.prepared_digest
            }
        };
        if !matches {
            return self.fail(GovernedFailureCode::InteractionConflict);
        }
        match resolution {
            InteractionResolution::Resolved { .. } => {
                self.last_interaction_resolution = Some(resolution);
                self.state = ReducerState::Prepared;
                GovernedAction::Authorize
            }
            InteractionResolution::Cancelled { .. } => {
                self.last_interaction_resolution = Some(resolution);
                self.state = ReducerState::Denied;
                GovernedAction::Observation(self.observation(ObservationOutcome::Rejected {
                    code: "interaction_cancelled".to_owned(),
                    details: None,
                }))
            }
        }
    }

    /// Reduces one Runtime execution fact after its required durable commit.
    pub fn apply_execution(&mut self, fact: ExecutionFact) -> GovernedAction {
        match (&self.state, fact) {
            (ReducerState::Authorized(grant), ExecutionFact::Started(_)) => {
                self.state = ReducerState::Started(grant.clone());
                GovernedAction::None
            }
            (ReducerState::Authorized(_), ExecutionFact::Unsupported { requirement }) => {
                if requirement.is_empty() {
                    return self.fail(GovernedFailureCode::CorruptRecoveryState);
                }
                self.fail(GovernedFailureCode::RequirementUnsupported)
            }
            (
                ReducerState::Started(grant),
                ExecutionFact::Completed {
                    receipt,
                    content,
                    truncated,
                },
            ) => {
                let Some(receipt) = receipt else {
                    return self.fail(GovernedFailureCode::CorruptRecoveryState);
                };
                if !self.receipt_binds(&receipt, grant, TerminalClassification::Completed)
                    || serde_json::to_vec(&content).map_or(true, |bytes| {
                        bytes.len() as u64 > grant.granted_requirements.max_output_bytes()
                    })
                {
                    return self.fail(GovernedFailureCode::CorruptRecoveryState);
                }
                self.state = ReducerState::Completed;
                GovernedAction::Observation(
                    self.observation(ObservationOutcome::Succeeded { content, truncated }),
                )
            }
            (
                ReducerState::Started(grant),
                ExecutionFact::Failed {
                    receipt,
                    code,
                    details,
                    partial,
                },
            ) => {
                let Some(receipt) = receipt else {
                    return self.fail(GovernedFailureCode::CorruptRecoveryState);
                };
                if code.is_empty()
                    || !self.receipt_binds(&receipt, grant, TerminalClassification::Failed)
                    || partial.as_ref().is_some_and(|value| {
                        serde_json::to_vec(value).map_or(true, |bytes| {
                            bytes.len() as u64 > grant.granted_requirements.max_output_bytes()
                        })
                    })
                {
                    return self.fail(GovernedFailureCode::CorruptRecoveryState);
                }
                self.state = ReducerState::Failed;
                GovernedAction::Observation(self.observation(ObservationOutcome::Failed {
                    code,
                    details,
                    partial,
                }))
            }
            (ReducerState::Started(_), ExecutionFact::Uncertain { evidence }) => {
                if evidence.is_empty() {
                    return self.fail(GovernedFailureCode::CorruptRecoveryState);
                }
                self.state = ReducerState::Uncertain;
                GovernedAction::Suspend(SuspensionRequirement::OperatorReconciliation { evidence })
            }
            _ => self.fail(GovernedFailureCode::InvocationConflict),
        }
    }

    fn grant_binds(&self, grant: &InvocationGrant) -> bool {
        grant.invocation_id == self.invocation_id
            && grant.prepared_digest == self.prepared.input_digest()
            && grant.tool_name == self.prepared.tool_name()
            && grant.tool_revision == self.prepared.tool_revision()
            && grant.granted_requirements.max_duration_ms()
                <= self.prepared.requirements().max_duration_ms()
            && grant.granted_requirements.max_output_bytes()
                <= self.prepared.requirements().max_output_bytes()
            && grant.granted_requirements.capabilities().all(|capability| {
                self.prepared
                    .requirements()
                    .capabilities()
                    .any(|requested| requested == capability)
            })
    }

    fn receipt_binds(
        &self,
        receipt: &EffectReceipt,
        grant: &InvocationGrant,
        terminal: TerminalClassification,
    ) -> bool {
        receipt.validate().is_ok()
            && receipt.invocation_id == self.invocation_id
            && receipt.prepared_digest == self.prepared.input_digest()
            && receipt.grant_id == grant.grant_id
            && receipt.terminal_classification == terminal
    }

    fn observation(&self, outcome: ObservationOutcome) -> GovernedObservation {
        GovernedObservation {
            invocation_id: self.invocation_id.clone(),
            prepared_digest: self.prepared.input_digest().to_owned(),
            model_call_id: self.prepared.model_call_id().to_owned(),
            tool_name: self.prepared.tool_name().to_owned(),
            outcome,
        }
    }

    fn fail(&mut self, code: GovernedFailureCode) -> GovernedAction {
        self.state = ReducerState::Failed;
        GovernedAction::Fail(GovernedEffectFailure { code })
    }
}

/// Selects recovery without treating a replay declaration as executor proof.
pub const fn recover_effect(
    position: RecoveryPosition,
    replay_class: ReplayClass,
    executor_proves_replay: bool,
) -> RecoveryDecision {
    match position {
        RecoveryPosition::Authorized => RecoveryDecision::RevalidateGrant,
        RecoveryPosition::ReceiptNoResult => RecoveryDecision::ReconstructFromReceipt,
        RecoveryPosition::Terminal => RecoveryDecision::ReturnTerminal,
        RecoveryPosition::StartedNoReceipt => match replay_class {
            ReplayClass::ReceiptRecoverable if executor_proves_replay => {
                RecoveryDecision::RecoverExecutorReceipt
            }
            ReplayClass::ReadOnly | ReplayClass::Idempotent if executor_proves_replay => {
                RecoveryDecision::RetrySameInvocation
            }
            _ => RecoveryDecision::ReconcileOperator,
        },
    }
}
